### PR TITLE
Addresses issue #38: Adding equals method to Coefficients, FixedEffectModel…

### PR DESCRIPTION
**Pull request by XianXing**
… and RandomEffectModel.

### CODE REVIEW

**Comment by XianXing** [comment]
@joshvfleming Can you please take a look at this PR? 

Please note that the I introduced a bug in https://github.com/linkedin/photon-ml/pull/36 by committing a file that was supposed to be committed in a separate PR (making the number files to be configurable when writing the random effect models, see more details in issue #20), which may cause the integration test to fail, and this PR will resolve this issue by reverting that change made in PR #36.

After this PR get merged, I will submit another PR (already prepared and ready) to address issue #20, as I promised to @ashelkovnykov long time ago.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverGameIntegTest.scala [hunk 1]
```diff
@@ -320,8 +320,7 @@ object DriverGameIntegTest {
     "validate-input-dirs" -> testPath,
     "feature-name-and-term-set-path" -> featurePath,
     "num-iterations" -> numIterations.toString,
-    "save-models-to-hdfs" -> true.toString,
-    "num-output-files-for-random-effect-model" -> "-1")
+    "save-models-to-hdfs" -> true.toString)
```
Comment (XianXing) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverGameIntegTest.scala]: This change will fix the bug introduced in #36.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/model/RandomEffectModel.scala [hunk 2]
```diff
@@ -93,6 +93,21 @@ protected[ml] class RandomEffectModel(
...
+  override def equals(that: Any): Boolean = {
+    that match {
+      case other: RandomEffectModel =>
+        val sameMetaData = this.randomEffectId == other.randomEffectId && this.featureShardId == other.featureShardId
+        val sameCoefficientsRDD = this.coefficientsRDD.fullOuterJoin(other.coefficientsRDD).mapPartitions(iterator =>
+            Iterator.single(iterator.forall { case (_, (coefficientsOpt1, coefficientsOpt2)) =>
+              coefficientsOpt1.isDefined && coefficientsOpt2.isDefined &&
+                  coefficientsOpt1.get.equals(coefficientsOpt2.get)
+            })
+        ).filter(sameCoefficients => !sameCoefficients).count() == 0
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/model/RandomEffectModel.scala]: Minor: you can use a point-free form here to make it a little more concise:

```scala
filter(! _)
```

**Comment by joshvfleming** [comment]
LGTM

**Comment by ashelkovnykov** [comment]
LGTM

File: photon-ml/src/main/scala/com/linkedin/photon/ml/model/RandomEffectModel.scala (hunk 2 above)
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/model/RandomEffectModel.scala]: Good point!

**Pull request merged**